model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java

```diff
@@ -279,4 +282,95 @@ protected void rollbackImpl() {
+        try {
+            if (getCache(offline) != null) {
+                existing = getCache(offline).putIfAbsent(key, session, lifespan, TimeUnit.MILLISECONDS, maxIdle, TimeUnit.MILLISECONDS);
+            }
+        } catch (RuntimeException exception) {
+            // debug level, it the import fail the transaction can continue with the data from the database.
+            LOG.debugf(exception, "Failed to import session %s", session);
+        }
+        if (existing == null) {
+            // keep track of the imported session for updates
+            updates.put(key, new SessionUpdatesList<>(realmModel, session));
+            return null;
+        }
+        updates.put(key, new SessionUpdatesList<>(realmModel, existing));
+        return existing;
+    }
+
+    /**
+     * Imports multiple sessions from an external source into the {@link Cache}.
+     * <p>
+     * If one or more sessions already exist in the {@link Cache}, or is expired, it will not be imported.
+     * <p>
+     * This transaction will keep track of further changes in the sessions.
+     *
+     * @param realmModel The {@link RealmModel} where the sessions belong to.
+     * @param sessions   The {@link Map} with the cache's key/session mapping to be imported.
+     * @param offline    {@code true} if it is an offline session.
+     */
+    public void importSessionsConcurrently(RealmModel realmModel, Map<K, SessionEntityWrapper<V>> sessions, boolean offline) {
+        var cache = getCache(offline);
+        if (sessions.isEmpty() || cache == null) {
+            //nothing to import
+            return;
+        }
+        var stage = CompletionStages.aggregateCompletionStage();
+        var allSessions = new ConcurrentHashMap<K, SessionEntityWrapper<V>>();
+        var updates = getUpdates(offline);
+        var lifespanFunction = getLifespanMsLoader(offline);
+        var maxIdleFunction = getMaxIdleMsLoader(offline);
+        sessions.forEach((key, session) -> {
+            if (updates.containsKey(key)) {
+                //nothing to import, already exists in transaction
+                return;
+            }
+            var clientModel = session.getClientIfNeeded(realmModel);
+            var sessionEntity = session.getEntity();
+            var lifespan = lifespanFunction.apply(realmModel, clientModel, sessionEntity);
+            var maxIdle = maxIdleFunction.apply(realmModel, clientModel, sessionEntity);
+            if (lifespan == SessionTimeouts.ENTRY_EXPIRED_FLAG || maxIdle == SessionTimeouts.ENTRY_EXPIRED_FLAG) {
+                //nothing to import, already expired
+                return;
+            }
+            var future = cache.putIfAbsentAsync(key, session, lifespan, TimeUnit.MILLISECONDS, maxIdle, TimeUnit.MILLISECONDS)
+                    .exceptionally(throwable -> {
+                        // debug level, it the import fail the transaction can continue with the data from the database.
```

Comment:
`// If the import fails, the transaction can continue with the data from the database.`